
<a href="https://colab.research.google.com/drive/1E6EBkVrgkr8iWlUB958b_VUyTnK9YdpA" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


### Árboles binarios

In [7]:
!pip install binarytree

Please see https://github.com/pypa/pip/issues/5599 for advice on fixing the underlying issue.
To avoid this problem you can invoke Python with '-m pip' instead of running pip directly.
/home/gibran/.local/lib/python3.8/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/home/gibran/.local/lib/python3.8/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
  Using cached setuptools-62.1.0-py3-none-any.whl (1.1 MB)
  Attempting uninstall: setuptools
    Found existing installation: setuptools 41.2.0
    Uninstalling setuptools-41.2.0:
      Successfully uninstalled setuptools-41.2.0
You should consider upgrading via the '/home/gibran/.pyenv/versions/3.8.2/bin/python3.8 -m pip install --upgrade pip' command.


In [15]:
from binarytree import Node
from binarytree import build
import numpy as np

In [9]:
root = Node(3)
root.left = Node(6)
root.right = Node(8)
 
# árbol binario
print('Árbol binario :', root)
 
# Lista de nodos
print('Lista de nodos :', list(root))
 
# recorrido inorder de nodos 
print('Inorder de nodos :', root.inorder)
 
# Propiedades de árboles
print('tamano de árbol :', root.size)
print('Altura de árbol :', root.height)
 
# Obten todas las propiedades
print('Propiedades del árbol : \n', root.properties)

Binary tree : 
  3
 / \
6   8

List of nodes : [Node(3), Node(6), Node(8)]
Inorder of nodes : [Node(6), Node(3), Node(8)]
Size of tree : 3
Height of tree : 1
Properties of tree : 
 {'height': 1, 'size': 3, 'is_max_heap': False, 'is_min_heap': True, 'is_perfect': True, 'is_strict': True, 'is_complete': True, 'leaf_count': 2, 'min_node_value': 3, 'max_node_value': 8, 'min_leaf_depth': 1, 'max_leaf_depth': 1, 'is_balanced': True, 'is_bst': False, 'is_symmetric': False}


In [13]:
nodes =[3, 6, 8, 2, 11, None, 13]
 
binary_tree = build(nodes)
print('Lista del árbolo binario :\n',
      binary_tree)
 
print('\nLista del árbol binario :',
      binary_tree.values)

Lista del árbolo binario :
 
    ___3
   /    \
  6      8
 / \      \
2   11     13


Lista del árbol binario : [3, 6, 8, 2, 11, None, 13]


### Expresiones regulares
1. Símbolo de escape:
    \
 2. Agrupación o captura:
    (...) (?:...) (?=...) (?!...) (?<=...) (?<!...)
 3. Cantidad (cuantificadores)
    a* a+ a? a{n, m}
 4. Secuencia y ancla
    abc ^ $ \b \B
 5. O (alternancia)
    a|b|c
 6. Átomos
    a [^abc] \t \r \n \d \D \s \S \w \W .



\ d Cualquier número; equivalente a [0-9]

 \ D Cualquier que no sea un dígito; equivalente a [^ 0-9]
 
 \ w Cualquier carácter literal; equivalente a [a-zA-Z0-9_]
 
 \ W Cualquier carácter no literal; equivalente a [^ a-zA-Z0-9_]
 
 \ s Cualquier espacio en blanco; equivalente a [\ f \ n \ r \ t \ v] (tenga en cuenta que el primer carácter entre corchetes es un símbolo de espacio)
 
 \ S Cualquier que no esté en blanco; equivalente a [^ \ f \ n \ r \ t \ v] (tenga en cuenta que inmediatamente después de ^ hay un símbolo de espacio)
 
 . Cualquier carácter excepto \ r \ n; equivalente a [^ \ r \ n]


In [1]:
import re

str = '<dl>(843) 542-4256</dl> <dl>(431) 270-9664</dl>'
 # Busque todos los caracteres en el formato "número-número" en la cadena, donde el primer número es de tres dígitos y el segundo número es de cuatro dígitos
pttn = r'\d\d\d\-\d\d\d\d'
print(re.findall(pttn, str))

['542-4256', '270-9664']


['542-4256', '270-9664']

In [3]:
str = 'never ever verb however everest'
pttn = r'er\b'
print(re.findall(pttn, str))
pttn = r'er\B'
print(re.findall(pttn, str))


['er', 'er', 'er']
['er', 'er']


^ Coincide con la posición inicial de la cadena buscada;

 $ Coincide con la posición final de la cadena buscada;
    
 \ b coincide con el límite de una palabra; er \ b puede coincidir con er en el codificador, pero no puede coincidir con er por error;

 \ B coincide con un límite sin palabras; er \ B puede coincidir con er por error, pero no en el codificador.


+ Significa que el átomo anterior debe aparecer al menos una vez, es decir: número de apariciones ≧ 1
 ? Indica que el átomo anterior solo puede aparecer una vez como máximo, es decir: 0 ≦ número de ocurrencias ≦ 1
 * Representa que el átomo anterior puede no aparecer, o aparecer una o más veces, es decir: número de apariciones ≧ 0
 {n} representa n veces que aparece el átomo anterior;
 {n,} significa que el átomo anterior aparece al menos n veces;
 {n, m} significa que el átomo anterior aparece al menos n veces y como máximo m veces

In [1]:
import re

str = 'never ever verb however everest'
pttn = r'er\b'
print(re.findall(pttn, str))
pttn = r'er\B'
print(re.findall(pttn, str))


['er', 'er', 'er']
['er', 'er']


In [4]:
# Match string
str = ['google', 'gooogle', 'goooogle', 'goooooogle']

 # Atraviesa la cuerda
for str_per in str:
         # o aparece al menos una vez 
    pttn = r'go+gle'
    print(re.findall(pttn, str_per))
    
         # o aparece como máximo una vez 
    pttn = r'go?gle'
    print(re.findall(pttn, str_per))

         # oNúmero de ocurrencias ≧ 0
    pttn = r'go?gle'
    print(re.findall(pttn, str_per))
    
         # o aparece al menos 2 veces y como máximo 5 veces
    pttn = r'go{2,5}gle'
    print(re.findall(pttn, str_per))
    
         # Advance: combinación simultánea de átomos
         # go o aparece al menos una vez
    pttn = r'[go]+gle'
    
         # go aparece al menos una vez
    pttn = r'(go)+gle'
    print(re.findall(pttn, str_per))

['google']
[]
[]
['google']
[]
['gooogle']
[]
[]
['gooogle']
[]
['goooogle']
[]
[]
['goooogle']
[]
['goooooogle']
[]
[]
[]
[]


1. Entendimiento del negocio: Determinar los objetivos del negocio, cuáles son sus necesidades y lo que se desea abarcar.

2. Entendimiento de los datos: De acuerdo a la fase anterior, con que datos cuento, que datos son útiles y cuáles necesito.

3. Preparación de datos: En esta fase se exploran los datos, se manipulan y se limpian, se hace la unión entre lo que quiere/necesita el negocio con los datos con los que cuenta la organización

4. Modelamiento: Teniendo en cuenta los puntos anteriores se realiza el análisis y se obtienen resultados.

5. Evaluación: Una vez se obtienen los resultados se evalúa el modelo teniendo en cuenta el objetivo del análisis

6. Despliegue: Una vez validado el modelo se realiza la implementación en producción, así como monitoreo y mantenimiento.

![python-ecosystem](https://miro.medium.com/max/1400/0*qc2yUKqzRGIfpICx.png)
<div style="text-align: center"> https://lauralpezb.medium.com/</div>

### Los fundamentos de las matrices NumPy

La manipulación de datos en Python es casi sinónimo de la manipulación de matrices NumPy: incluso las herramientas más nuevas, como Pandas (Capítulo 3), se construyen alrededor de la matriz NumPy. Esta sección presentará varios ejemplos del uso de la manipulación de arreglos NumPy para acceder a datos y subarreglos, y para dividir, remodelar y unir los arreglos. Si bien los tipos de operaciones que se muestran aquí pueden parecer un poco áridos y pedantes, comprenden los componentes básicos de muchos otros ejemplos que se usan a lo largo del libro. ¡Conócelos bien!

Cubriremos algunas categorías de manipulaciones básicas de matrices aquí:

Atributos de los arreglos: determinación del tamaño, la forma, el consumo de memoria y los tipos de datos de los arreglos
Indexación de arreglos: Obtener y establecer el valor de elementos de arreglo individuales
Segmentación de arreglos: obtención y configuración de subarreglos más pequeños dentro de un arreglo más grande
Remodelación de arreglos: Cambiar la forma de un arreglo dado
Unión y división de arreglos: combinación de múltiples arreglos en uno y división de un arreglo en muchos

In [17]:
np.random.seed(0)

x1 = np.random.randint(10, size=6)  # dim 1
x2 = np.random.randint(10, size=(3, 4))  # dim 2
x3 = np.random.randint(10, size=(3, 4, 5))  # dim 3

In [18]:
x1, x2, x3

(array([5, 0, 3, 3, 7, 9]),
 array([[3, 5, 2, 4],
        [7, 6, 8, 8],
        [1, 6, 7, 7]]),
 array([[[8, 1, 5, 9, 8],
         [9, 4, 3, 0, 3],
         [5, 0, 2, 3, 8],
         [1, 3, 3, 3, 7]],
 
        [[0, 1, 9, 9, 0],
         [4, 7, 3, 2, 7],
         [2, 0, 0, 4, 5],
         [5, 6, 8, 4, 1]],
 
        [[4, 9, 8, 1, 1],
         [7, 9, 9, 3, 6],
         [7, 2, 0, 3, 5],
         [9, 4, 4, 6, 4]]]))

In [19]:
print("x3 ndim: ", x3.ndim)
print("x3 shape:", x3.shape)
print("x3 size: ", x3.size)

x3 ndim:  3
x3 shape: (3, 4, 5)
x3 size:  60


In [20]:
x1, x1[0], x2[0, 0]

(array([5, 0, 3, 3, 7, 9]), 5, 3)

In [21]:
x2[0, 0] = 12

In [23]:
#x[start:stop:step]

x1[:5], x1[::2], x1[1::2]

(array([5, 0, 3, 3, 7]), array([5, 3, 7]), array([0, 3, 9]))

Crando copias 

In [24]:
x2_sub_copy = x2[:2, :2].copy()
print(x2_sub_copy)

[[12  5]
 [ 7  6]]


In [25]:
grid = np.arange(1, 10).reshape((3, 3))
print(grid)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


Concatenar y Dividir arreglos

In [27]:
x = np.array([1, 2, 3])
y = np.array([3, 2, 1])
np.concatenate([x, y])

array([1, 2, 3, 3, 2, 1])

In [28]:
x = [1, 2, 3, 99, 99, 3, 2, 1]
x1, x2, x3 = np.split(x, [3, 5])
print(x1, x2, x3)

[1 2 3] [99 99] [3 2 1]


### Cálculo en matrices NumPy: funciones universales

Hasta ahora, hemos discutido algunos de los aspectos básicos de NumPy; En las próximas secciones, profundizaremos en las razones por las que NumPy es tan importante en el mundo de la ciencia de datos de Python. Es decir, proporciona una interfaz fácil y flexible para el cálculo optimizado con matrices de datos.

El cálculo en matrices NumPy puede ser muy rápido o muy lento. La clave para hacerlo rápido es usar operaciones vectorizadas, generalmente implementadas a través de las funciones universales de NumPy (ufuncs). Esta sección motiva la necesidad de los ufuncs de NumPy, que se pueden usar para hacer que los cálculos repetidos en los elementos de la matriz sean mucho más eficientes. Luego presenta muchos de los ufuncs aritméticos más comunes y útiles disponibles en el paquete NumPy.

La relativa lentitud de Python generalmente se manifiesta en situaciones en las que se repiten muchas operaciones pequeñas, por ejemplo, recorriendo matrices para operar en cada elemento. Por ejemplo, imagina que tenemos una matriz de valores y nos gustaría calcular el recíproco de cada uno. Un enfoque sencillo podría verse así:

In [29]:
def compute_reciprocals(values):
    output = np.empty(len(values))
    for i in range(len(values)):
        output[i] = 1.0 / values[i]
    return output
        
values = np.random.randint(1, 10, size=5)
compute_reciprocals(values)

array([0.2       , 0.25      , 0.2       , 0.2       , 0.11111111])

Operación sobre arreglos

In [33]:
x = np.arange(4)
print("x     =", x)
print("x + 5 =", x + 5)
print("x - 5 =", x - 5)
print("x * 2 =", x * 2)
print("x / 2 =", x / 2)
print("x // 2 =", x // 2)
print("-x     = ", -x)
print("x ** 2 = ", x ** 2)
print("x % 2  = ", x % 2)

x     = [0 1 2 3]
x + 5 = [5 6 7 8]
x - 5 = [-5 -4 -3 -2]
x * 2 = [0 2 4 6]
x / 2 = [0.  0.5 1.  1.5]
x // 2 = [0 0 1 1]
-x     =  [ 0 -1 -2 -3]
x ** 2 =  [0 1 4 9]
x % 2  =  [0 1 0 1]


Tabla con operaciones en NumPy:

| Operator	    | Equivalent ufunc    | Description                           |
|---------------|---------------------|---------------------------------------|
|``+``          |``np.add``           |Addition (e.g., ``1 + 1 = 2``)         |
|``-``          |``np.subtract``      |Subtraction (e.g., ``3 - 2 = 1``)      |
|``-``          |``np.negative``      |Unary negation (e.g., ``-2``)          |
|``*``          |``np.multiply``      |Multiplication (e.g., ``2 * 3 = 6``)   |
|``/``          |``np.divide``        |Division (e.g., ``3 / 2 = 1.5``)       |
|``//``         |``np.floor_divide``  |Floor division (e.g., ``3 // 2 = 1``)  |
|``**``         |``np.power``         |Exponentiation (e.g., ``2 ** 3 = 8``)  |
|``%``          |``np.mod``           |Modulus/remainder (e.g., ``9 % 4 = 1``)|
